helix-rest/src/main/java/org/apache/helix/rest/common/HelixDataAccessorWrapper.java

```diff
@@ -1,24 +1,158 @@
+
+import org.apache.helix.HelixProperty;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixDataAccessor;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.RESTConfig;
+import org.apache.helix.rest.client.CustomRestClient;
+import org.apache.helix.rest.client.CustomRestClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This is a wrapper for {@link ZKHelixDataAccessor} that caches the result of the batch reads it
  * performs.
  * Note that the usage of this object is valid for one REST request.
  */
 public class HelixDataAccessorWrapper extends ZKHelixDataAccessor {
+  private static final Logger LOG = LoggerFactory.getLogger(HelixDataAccessorWrapper.class);
+  private static final ExecutorService POOL = Executors.newCachedThreadPool();
+
+  private static final String PARTITION_HEALTH_KEY = "PARTITION_HEALTH";
+  private static final String IS_HEALTHY_KEY = "IS_HEALTHY";
+  private static final String EXPIRY_KEY = "EXPIRE";
+
+  private final CustomRestClient _restClient;
   private final Map<PropertyKey, HelixProperty> _propertyCache = new HashMap<>();
   private final Map<PropertyKey, List<String>> _batchNameCache = new HashMap<>();
 
   public HelixDataAccessorWrapper(ZKHelixDataAccessor dataAccessor) {
     super(dataAccessor);
+    _restClient = CustomRestClientFactory.get();
+  }
+
+  HelixDataAccessorWrapper(ZKHelixDataAccessor dataAccessor, CustomRestClient client) {
+    super(dataAccessor);
+    _restClient = client;
+  }
+
+  public List<ExternalView> getExternalViews() {
+    return getChildNames(keyBuilder().externalViews()).stream()
+        .map(externalView -> (ExternalView) getProperty(keyBuilder().externalView(externalView)))
+        .collect(Collectors.toList());
+  }
+
+  public Map<String, Map<String, Boolean>> getPartitionHealthOfInstance(RESTConfig restConfig,
+      Map<String, String> customPayLoads) {
+    // Only checks the instances are online with valid reports
+    List<String> liveInstances = getChildNames(keyBuilder().liveInstances());
+    // Make a parallel batch call for getting all healthreports from ZK.
+    List<HelixProperty> zkHealthReports = getProperty(liveInstances.stream()
+        .map(instance -> keyBuilder().healthReport(instance, PARTITION_HEALTH_KEY))
+        .collect(Collectors.toList()), false);
+    Map<String, Future<Map<String, Boolean>>> parallelTasks = new HashMap<>();
+    for (int i = 0; i < liveInstances.size(); i++) {
+      String liveInstance = liveInstances.get(i);
+      Optional<ZNRecord> maybeHealthRecord =
+          Optional.ofNullable(zkHealthReports.get(i)).map(HelixProperty::getRecord);
+      parallelTasks.put(liveInstance, POOL.submit(() -> {
+        if (maybeHealthRecord.isPresent()) {
+          return getPartitionsHealth(liveInstance, maybeHealthRecord.get(), restConfig,
+              customPayLoads);
+        } else {
```

Comment:
else clause is not needed.

Reply:
It's different than the previous implementation. The private method `getPartitionsHealth` only submits queries when it's expired. The else case deals with when we don't have any record for the instance, it needs to submit a query of empty partitions list.

Reply:
I am not saying totally remove the logic but just coding style. Since you change it in java 8, I will review it again.